### PR TITLE
Simulator with large range inputs

### PR DESCRIPTION
**Pull request by ckyiricky**
This branch is derived from branch simulator-tests, so compared to master, it could have some overlapped changes with the tests PR. 
[Add] Add range input generator
    Add range input generator for larger scale tests
    Add shell script to run all algorithms
    Update one algorithm name in algorithms.go to be consistent with others
[Update] Update process to multi-threaded
    Update process to multi-threaded which can process large data input files
    Update local-shared algorithm that only zones with local SG > 1 endpoint
    can be candidates(give endpoints to others)
    Update local-shared algorithm that urgent zones will no longer be added
    to 'needed list' separately
[Add] Add result parser 
[Update] Optimize local-shared algorithm with corner cases.
[Update] Evaluation mechanism with only positive deviation

### CODE REVIEW

File: modeling/algorithm/local-shared-algorithm.go [hunk 1]
```diff
@@ -225,3 +228,9 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
 	weightedComp.Weight = weight
 	sliceGroup.Composition[zone] = weightedComp
 }
+
+// detect whether a zone is valid to contribute endpoints to other zones
+func validContributor(deviation float64, zoneName string, sliceGroup types.EndpointSliceGroup) bool {
+	// sliceGroup.Composition[zoneName] is guaranteed to be non-nil by caller
+	return deviation > 0 && sliceGroup.Composition[zoneName].Number > 1
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: I'd feel better if there were a nil check on `sliceGroup.Composition[zoneName]` here since this can be called from multiple places.

File: process/input-parser.go [hunk 2]
```diff
@@ -28,13 +28,14 @@ import (
 	"k8s.io/klog/v2"
 )
 
-// parseInput parses input csv file to instances of inputData and returns a
-// slice of them
-func parseInput(file string) ([]inputData, error) {
+// paserInput parses input csv file to instances of inputData and puts them into
```
Comment (robscott) [inline process/input-parser.go]:
```suggestion
// parseInput parses an input csv file to instances of inputData and puts them into
```

File: process/input-parser.go [hunk 3]
```diff
@@ -47,21 +48,33 @@ func parseInput(file string) ([]inputData, error) {
...
+			}
+		}()
+
+		for {
+			var data inputData
+			var done bool
+			data, done, err = readOneRow(zoneNames, reader)
+			if done {
+				break
+			}
```
Comment (robscott) [inline process/input-parser.go]: Would something like this work?
```suggestion
		for data, done, err := readOneRow(zoneNames, reader); !done {
```

File: process/output-parser.go [hunk 4]
```diff
@@ -27,13 +27,16 @@ import (
...
 	if err != nil {
 		return err
 	}
 	defer func() {
 		cerr := outputFile.Close()
+		if cerr != nil {
+			klog.Errorf("close output file %s with an error %v", file, cerr)
+		}
 		if err == nil {
 			err = cerr
```
Comment (robscott) [inline process/output-parser.go:41]: maybe this should just be `return cerr`

File: process/output-parser.go [hunk 5]
```diff
@@ -48,7 +51,11 @@ func parseResult(file string, outputArray []outputData) (err error) {
 		return err
 	}
 
-	for _, rowData := range outputArray {
+	for {
+		rowData, more := <-outputQueue
+		if !more {
+			break
+		}
```
Comment (robscott) [inline process/output-parser.go]: Would this work?
```suggestion
	for rowData, more := <-outputQueue; !more {
```

File: process/process.go [hunk 6]
```diff
@@ -66,33 +71,49 @@ type outputData struct {
...
-	for _, rowData := range inputArray {
-		err := model.UpdateRegion(rowData.zones)
+// outputData structure and puts them in a queue(channel)
+func startSimulation(algName string, inputQueue <-chan inputData) (<-chan outputData, error) {
+	outputQueue := make(chan outputData)
+	errC := make(chan error, 1)
+	defer close(errC)
+
+	go func() {
+		defer close(outputQueue)
```
Comment (robscott) [inline process/process.go:87]: Why is this being closed inside the goroutine while the error channel is closed outside of it?

File: process/process.go [hunk 7]
```diff
@@ -66,33 +71,49 @@ type outputData struct {
...
-		simRes, err := model.StartSimulation()
-		if err != nil {
-			return outputArray, fmt.Errorf("error starting simulation for input : %s, %v", rowData.name, err)
+		errC <- nil
+
+		for {
+			rowData, more := <-inputQueue
+			if !more {
+				break
+			}
```
Comment (robscott) [inline process/process.go]:
```suggestion
		for rowData, more := <-inputQueue; !more {
```

File: process/process.go [hunk 8]
```diff
@@ -66,33 +71,49 @@ type outputData struct {
...
+			simRes, err := model.StartSimulation()
+			if err != nil {
+				klog.Errorf("error starting simulation for input : %s, %v", rowData.name, err)
+				continue
+			}
+			outputQueue <- outputData{name: rowData.name,
+				endpoints:      model.GetNumberOfEndpoints(),
+				endpointSlices: model.GetNumberOfEndpointSlices(),
+				result:         simRes}
```
Comment (robscott) [inline process/process.go]:
```suggestion
			oData, err := runSimulation(model, rowData.Zones)
			if err == nil {
				outputQueue <- oData
			}
```

File: process/range-input-generator.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+"""
```
Comment (robscott) [inline process/range-input-generator.py]: I think this should be a separate dir away from the go code.

File: run-all.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2020 Google LLC
```
Comment (robscott) [inline run-all.sh:2]: Since both this and the python script are a new way to interact with this project, this PR should also include some docs updates.

**Review by robscott — COMMENTED** [review]
Thanks for the work on this! This mostly LGTM, still need to dig into the python script though.

File: run-all.sh (hunk 10 above)
Comment (robscott) [reply run-all.sh:2]: This file could use `chmod +x`

File: run-all.sh [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+declare -a algs=("SharedGlobal" "SharedMultiZone" "Local" "LocalWeighted" "LocalOpt" "LocalShared" "Original")
+
+for alg in ${algs[@]}; do
+    echo "Running $alg"
+    go run main.go -input=example/range-input.csv -alg=$alg -output=example/$alg-range-output.csv
```
Comment (robscott) [inline run-all.sh]: `open example/range-input.csv: no such file or directory`

File: process/range-input-generator.py (hunk 9 above)
Comment (robscott) [reply process/range-input-generator.py]: Also should add a note somewhere that this requires python 3.

File: process/range-input-generator.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import csv
+import os
+from itertools import combinations 
+
+# Print iterations progress
+def printProgressBar (iteration, total, progress="Progress", length = 100, fill = '█', printEnd = "\r"):
```
Comment (robscott) [inline process/range-input-generator.py]: As much as I personally prefer camel case, snake case is the standard for Python. https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: process/range-input-generator.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    writer = csv.DictWriter(csv_file, fieldnames=field_names)
+
+    writer.writeheader()
+    # generate range input: 3 zones
+    # nodes are the conbinations from 1-10
+    # endpoints range from 0-100 with step = 1
+    nodes_comb = list(combinations(range(1, 11), 3))
+    for index, nodes in enumerate(nodes_comb):
+        generateOneSection(writer, nodes, suffix='-'+str(nodes), bar=False)
+        printProgressBar(index+1, len(nodes_comb), progress="nodes iteration")
```
Comment (robscott) [inline process/range-input-generator.py]: This ends up feeling a bit inconsistent with output for the other section. Seems like the names should be more similar (0-100 endpoints, 100-1000 endpoints) and ideally the progress bars would line up more closely. Overall this is a really cool interface though.

```
nodes iteration: |████████████████████████████████████████████████████████████████████████████████████████████████████| 100.00% Complete
100-1000 endpoints [30, 30, 30] nodes, step: 7: |██████████████████████████████████████████████████████████████████████| 100.00% Complete
```

File: process/range-input-generator.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+    # generate range input: 3 zones
+    # nodes = 30 for every zone
+    # endpoints range from 100-1000 with step = 7
+    generateOneSection(writer, ep_range=[100, 1001], step = 7, suffix='-high')
```
Comment (robscott) [inline process/range-input-generator.py]: I just looked at the first and last 100 lines of this output and it seems like the deduping might be a little too eager. As an example, here's the last lines of my output:

```
349494-high,30 968,30 975,30 982
349495-high,30 968,30 975,30 989
349496-high,30 968,30 975,30 996
349497-high,30 968,30 982,30 989
349498-high,30 968,30 982,30 996
349499-high,30 968,30 989,30 996
349500-high,30 975,30 982,30 989
349501-high,30 975,30 982,30 996
349502-high,30 975,30 989,30 996
349503-high,30 982,30 989,30 996
```

Maybe this is covered by another test case elsewhere, but it seems like it would also be useful to test examples where some zones had the same number of endpoints like 996, 996, 989.

File: run-all.sh (hunk 10 above)
Comment (robscott) [reply run-all.sh:2]: I think this file should also start with 

```
#!/bin/bash
```

File: run-all.sh (hunk 11 above)
Comment (robscott) [reply run-all.sh]: Probably want to add a check with bash for the existence of this file before starting this loop. If it doesn't exist you could exit and log that the python generation script needs to run first.

Comment (robscott) [reply run-all.sh]: Also as a sidebar, would probably be good to use a directory other than `example` here and/or add `range-input.csv` to `.gitignore`

**Review by robscott — COMMENTED** [review]
Thanks for all the work on this! Got a chance to run the python and shell scripts and both work well on my computer as well.

File: modeling/algorithm/local-shared-algorithm.go (hunk 1 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: added nil check

File: process/input-parser.go (hunk 2 above)
Comment (ckyiricky) [reply process/input-parser.go]: changed

File: process/input-parser.go (hunk 3 above)
Comment (ckyiricky) [reply process/input-parser.go]: changed to an embedded way: 
`for data, done, rerr := readOneRow(zoneNames, reader); !done; data, done, rerr = readOneRow(zoneNames, reader) {`
Need to call it again in the 3rd section and use another name other than err to avoid shadow previous err variable.

File: process/output-parser.go (hunk 5 above)
Comment (ckyiricky) [reply process/output-parser.go]: similar to above, changed to an embedded way.

File: process/process.go (hunk 6 above)
Comment (ckyiricky) [reply process/process.go:87]: changed

File: process/process.go (hunk 7 above)
Comment (ckyiricky) [reply process/process.go]: changed as the other two places.

File: process/process.go (hunk 8 above)
Comment (ckyiricky) [reply process/process.go]: done

File: run-all.sh (hunk 10 above)
Comment (ckyiricky) [reply run-all.sh:2]: TODO: doc update in future PRs

File: run-all.sh (hunk 11 above)
Comment (ckyiricky) [reply run-all.sh]: added

File: process/range-input-generator.py (hunk 12 above)
Comment (ckyiricky) [reply process/range-input-generator.py]: changed

File: process/range-input-generator.py (hunk 13 above)
Comment (ckyiricky) [reply process/range-input-generator.py]: changed

File: process/range-input-generator.py (hunk 14 above)
Comment (ckyiricky) [reply process/range-input-generator.py]: fixed, need further discussion though

**Comment by ckyiricky** [comment]
Add a result parser for large output
Change some coding style

**Comment by ckyiricky** [comment]
[Update] Optimize local-shared algorithm with corner cases
    Update local-shared algorithm to maintain max deviation less than a
    threshold
    Update local-shared test with more corner cases
    Update evaluation mechanism to ignore negative deviation(less traffic
    than expected)
    Update readme for large dataset input usage

File: README.md [hunk 15]
```diff
@@ -12,13 +12,33 @@ Kubernetes. This specific project will include:
...
 
 example of intput file (csv): each zone with number of nodes first, number of endpoints next
 ```
 input name, zone1, zone2, zone3  
 perfect input, 10 10, 10 10, 20 20
 ```
+### Multiple algorithms usage
+`sh ./run-all.sh [input-file]`
+
+input-file is optional, ./data/range-input.csv is set as default.
```
Comment (robscott) [inline README.md]: Maybe note that this file can be generated by `range-input-generator.py` described below.

File: modeling/algorithm/local-shared-algorithm.go [hunk 16]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
-	if sliceGroup.Composition == nil {
+	if sliceGroups[zoneName].Composition == nil || sliceGroups[zoneName].NumberOfEndpoints() == 1 {
 		return false
 	}
-	return deviation > 0 && sliceGroup.Composition[zoneName].Number > 1
+	return !alg.invalidDeviation(zoneName, region, sliceGroups, -1)
+}
+
+// check if endpoints in receiveZone have invalid deviation
+func (alg LocalSharedSliceAlgorithm) invalidDeviation(receiveZone string, region types.RegionInfo, sliceGroups map[string]types.EndpointSliceGroup, delta int) bool {
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: Maybe something like `deviationAboveThreshold()`?

File: modeling/algorithm/local-shared-algorithm.go [hunk 17]
```diff
@@ -114,11 +118,10 @@ func (alg LocalSharedSliceAlgorithm) CreateSliceGroups(region types.RegionInfo)
 	availablePool.SliceGroups = sliceGroups
 	receiverPool.SliceGroups = sliceGroups
 
-	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, sliceGroups, &availablePool, &receiverPool)
+	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, region, sliceGroups, &availablePool, &receiverPool)
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go:121]: Do we need both success and err here? If so we should handle err before success.

File: modeling/algorithm/local-shared-algorithm.go [hunk 18]
```diff
@@ -147,7 +149,15 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
 	}
 	// take the ceil, if the deviation > 0, we have to make sure there is at
 	// least one endpoint in this shared SG (avoid 0.x ending up with 0)
-	mergedED.deviation = int(math.Ceil(mergedDeviation))
+	if mergedDeviation >= 1 {
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: sum of expectedEndpoints?

File: modeling/algorithm/local-shared-algorithm.go [hunk 19]
```diff
@@ -193,7 +207,7 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
 		candidate := heap.Pop(availablePool).(string)
 		// candidate is guaranteed to have a local owned SG, omit the second
 		// returned value
-		deviation, _ := getEndpointsDeviation(availablePool.Region, availablePool.SliceGroups, candidate)
+		deviation, _ := getEndpointsDeviation(region, sliceGroups, candidate)
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: Are we using the second return value anywhere?

File: modeling/algorithm/local-shared-algorithm.go [hunk 20]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+// helper function helps to keep all the endpoints with a traffic load deviation
+// less than threshold, return false if it can't.
+func (alg LocalSharedSliceAlgorithm) keepDeviationBelowThreshold(availablePool *ZonePriorityQueue, receiverPool *ZonePriorityQueue) bool {
+	region := availablePool.Region
+	sliceGroups := availablePool.SliceGroups
+	// get zones with deviation >= threshold
+	var urgentZones []string
+	for receiverPool.Len() > 0 {
+		receiveZone := receiverPool.ZoneNames[0]
+		if !alg.invalidDeviation(receiveZone, region, sliceGroups, 0) {
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: `deviationAboveThreshold()`

Comment (robscott) [reply modeling/algorithm/local-shared-algorithm.go]: Maybe explain why you're breaking here?

File: modeling/algorithm/local-shared-algorithm.go [hunk 21]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+			break
+		}
+		urgentZones = append(urgentZones, receiveZone)
+		heap.Pop(receiverPool)
+	}
+
+	// get extra endpoints from zones
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: Maybe add a bit more of an explanation of what extra endpoints are

File: modeling/algorithm/local-shared-algorithm.go [hunk 22]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+	extraEndpoints := map[string]int{}
+	extraEndpointsNumber := 0
+	if len(urgentZones) > 0 {
+		// if extraEndpointsNumber >= len(urgentZones), we just assign one
+		// enpoint to each urgent zone, their deviation will go below threshold
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]:
```suggestion
		// endpoint to each urgent zone, their deviation will go below threshold
```

File: modeling/algorithm/local-shared-algorithm.go [hunk 23]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+func (alg LocalSharedSliceAlgorithm) getMoreEndpoints(availablePool *ZonePriorityQueue, extraEndpoints map[string]int, urgentZones []string) bool {
+	sliceGroups := availablePool.SliceGroups
+	region := availablePool.Region
+	extraEndpointsNumber := 0
+	for _, num := range extraEndpoints {
+		extraEndpointsNumber += num
+	}
+	// continuously check if the extra endpoints can help build a valid shared
+	// sliceGroup
+	for !alg.trafficDeviationAdequate(urgentZones, availablePool.Region, sliceGroups, extraEndpointsNumber) {
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: Function is asking if there are enough extra endpoints for a shared slice to get deviation below threshold. Maybe a function name like `enoughExtraEndpoints` or `sufficientExtraEndpoints` would be clearer.

File: modeling/algorithm/local-shared-algorithm.go [hunk 24]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+// deviation less than threshold, it gets more endpoints from other zones.
+func (alg LocalSharedSliceAlgorithm) getMoreEndpoints(availablePool *ZonePriorityQueue, extraEndpoints map[string]int, urgentZones []string) bool {
+	sliceGroups := availablePool.SliceGroups
+	region := availablePool.Region
+	extraEndpointsNumber := 0
+	for _, num := range extraEndpoints {
+		extraEndpointsNumber += num
+	}
+	// continuously check if the extra endpoints can help build a valid shared
+	// sliceGroup
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]:
```suggestion
	// rebalance endpoints until deviations are below threshold or nothing left in availablePool
```

File: modeling/algorithm/local-shared-algorithm.go [hunk 25]
```diff
@@ -147,7 +149,15 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
...
 	// least one endpoint in this shared SG (avoid 0.x ending up with 0)
-	mergedED.deviation = int(math.Ceil(mergedDeviation))
+	if mergedDeviation >= 1 {
+		// round mergedDeviation to 3 decimal places to deal with internal float
+		// precision lose, i.e. 1.5 will end up with 1.49999998 that be rounded
+		// to 1 instead of 2
+		mergedDeviation = math.Ceil(mergedDeviation*1000) / 1000
+		mergedED.deviation = int(math.Round(mergedDeviation))
+	} else {
+		mergedED.deviation = int(math.Ceil(mergedDeviation))
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: Ceiling for 1.0000001 probably doesn't make sense

File: modeling/algorithm/local-shared-algorithm.go [hunk 26]
```diff
@@ -147,7 +149,15 @@ func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpoin
 	}
 	// take the ceil, if the deviation > 0, we have to make sure there is at
 	// least one endpoint in this shared SG (avoid 0.x ending up with 0)
-	mergedED.deviation = int(math.Ceil(mergedDeviation))
+	if mergedDeviation >= 1 {
+		// round mergedDeviation to 3 decimal places to deal with internal float
+		// precision lose, i.e. 1.5 will end up with 1.49999998 that be rounded
+		// to 1 instead of 2
+		mergedDeviation = math.Ceil(mergedDeviation*1000) / 1000
+		mergedED.deviation = int(math.Round(mergedDeviation))
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: I don't think this makes sense here but it may make more sense below for ceil.

File: modeling/algorithm/local-shared-algorithm.go [hunk 27]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+	sharedSG.Label = sharedLabel
+	sliceGroups[sharedLabel] = sharedSG
+}
+
+// when it failed to create a shared sliceGroup for urgent zones to achieve
+// deviation less than threshold, it gets more endpoints from other zones.
+func (alg LocalSharedSliceAlgorithm) getMoreEndpoints(availablePool *ZonePriorityQueue, extraEndpoints map[string]int, urgentZones []string) bool {
+	sliceGroups := availablePool.SliceGroups
+	region := availablePool.Region
+	extraEndpointsNumber := 0
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go:414]: Add a comment to describe how this is used.

File: modeling/algorithm/local-shared-algorithm.go [hunk 28]
```diff
@@ -229,11 +245,163 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+	for zone, number := range extraEndpoints {
+		updateSGComposition(sharedSG, zone, number, 1)
+	}
+	sharedSG.Label = sharedLabel
+	sliceGroups[sharedLabel] = sharedSG
+}
+
+// when it failed to create a shared sliceGroup for urgent zones to achieve
+// deviation less than threshold, it gets more endpoints from other zones.
+func (alg LocalSharedSliceAlgorithm) getMoreEndpoints(availablePool *ZonePriorityQueue, extraEndpoints map[string]int, urgentZones []string) bool {
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: This is a hard one to name, but maybe one of the following would work:

- `getExtraEndpoints`
- `getSharedSliceEndpoints`
- `getEndpointsForSharedSlice`
- `getExtraEndpointsForSharedSlice`

File: modeling/algorithm/local-shared-algorithm.go [hunk 29]
```diff
@@ -225,3 +244,164 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+	}
+	for zone, number := range extraEndpoints {
+		updateSGComposition(sharedSG, zone, number, 1)
+	}
+	sharedSG.Label = sharedLabel
+	sliceGroups[sharedLabel] = sharedSG
+}
+
+// when it failed to create a shared sliceGroup for urgent zones to achieve
+// deviation less than threshold, it gets more endpoints from other zones.
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]:
```suggestion
// [function name] attempts to get extra endpoints that could be used for a shared slice to get deviation below threshold. Returns false if it can't.
```

File: modeling/algorithm/local-shared-algorithm.go [hunk 30]
```diff
@@ -225,3 +244,164 @@ func updateSGComposition(sliceGroup types.EndpointSliceGroup, zone string, delta
...
+		sharedLabel += fmt.Sprintf("-%s", urgentZone)
+		for zone, contribution := range sliceGroups[urgentZone].Composition {
+			updateSGComposition(sharedSG, zone, contribution.Number, contribution.Weight)
+		}
+		sharedSG.ZoneTrafficWeights[urgentZone] = 1
+		delete(sliceGroups, urgentZone)
+	}
+	for zone, number := range extraEndpoints {
+		updateSGComposition(sharedSG, zone, number, 1)
+	}
```
Comment (robscott) [inline modeling/algorithm/local-shared-algorithm.go]: Some comments would be helpful here to explain what `updateSGComposition` is responsible for here. Is it taking endpoints previously allocated to the zone's slice group and giving them to the shared slice group?

File: data/range-input-generator.py [hunk 31]
```diff
@@ -0,0 +1,69 @@
+#! /usr/bin/env python3
```
Comment (robscott) [inline data/range-input-generator.py]: I think it would good to separate the scripts to generate data from where the data is stored. Maybe these scripts could move into a new `scripts` or `hack` director.

**Review by robscott — COMMENTED** [review]
Thanks for your work on this! Some great improvements here.

File: README.md (hunk 15 above)
Comment (ckyiricky) [reply README.md]: added

File: modeling/algorithm/local-shared-algorithm.go (hunk 16 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: done

File: modeling/algorithm/local-shared-algorithm.go (hunk 17 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go:121]: handle the error check at first.

File: modeling/algorithm/local-shared-algorithm.go (hunk 18 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: changed

File: modeling/algorithm/local-shared-algorithm.go (hunk 19 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: added usage of the second value, which should always be true, if its not, return error (unexpected)

File: modeling/algorithm/local-shared-algorithm.go (hunk 20 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: added

File: modeling/algorithm/local-shared-algorithm.go (hunk 21 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: added

File: modeling/algorithm/local-shared-algorithm.go (hunk 22 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: fixed

File: modeling/algorithm/local-shared-algorithm.go (hunk 23 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: changed to sufficientExtraEndpointsForSharedSlice

File: modeling/algorithm/local-shared-algorithm.go (hunk 24 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: changed

File: modeling/algorithm/local-shared-algorithm.go (hunk 27 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go:414]: added

File: modeling/algorithm/local-shared-algorithm.go (hunk 28 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: changed to `getExtraEndpointsForSharedSlice`

File: modeling/algorithm/local-shared-algorithm.go (hunk 29 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: changed

File: modeling/algorithm/local-shared-algorithm.go (hunk 30 above)
Comment (ckyiricky) [reply modeling/algorithm/local-shared-algorithm.go]: added

File: data/range-input-generator.py (hunk 31 above)
Comment (ckyiricky) [reply data/range-input-generator.py]: moved to hack dir

**Comment by ckyiricky** [comment]
Update function names
Move scripts to hack dir
Add more comments

File: process/input-parser.go [hunk 32]
```diff
@@ -28,13 +28,14 @@ import (
 	"k8s.io/klog/v2"
 )
 
-// parseInput parses input csv file to instances of inputData and returns a
-// slice of them
-func parseInput(file string) ([]inputData, error) {
+// paserInput parses an input csv file to instances of inputData and puts them
```
Comment (robscott) [inline process/input-parser.go]:
```suggestion
// parseInput parses an input csv file to instances of inputData and puts them
```

File: process/process.go [hunk 33]
```diff
@@ -31,18 +30,24 @@ const inZoneTrafficScoreWeight, deviationScoreWeight, sliceScoreWeight = 0.4, 0.
 // StartProcessing starts parsing input file, running simulation and
 // generating output file
 func StartProcessing(inputFile string, outputFile string, alg string) error {
-	inputArray, err := parseInput(inputFile)
+
+	// initialize a goroutine to keep reading row data from input file and
+	// putting the converted row data into a queue
```
Comment (robscott) [inline process/process.go]:
```suggestion
	// initialize a goroutine to keep read row data from input file and
	// put the converted row data into a queue
```

File: process/process.go [hunk 34]
```diff
@@ -31,18 +30,24 @@ const inZoneTrafficScoreWeight, deviationScoreWeight, sliceScoreWeight = 0.4, 0.
...
+	inputQueue, err := parseInput(inputFile)
 	if err != nil {
 		return err
 	}
 
-	outputArray, err := startSimulation(alg, inputArray)
+	// initialize a goroutine to keep processing row data from inputQueue and
+	// putting the processed data into another queue
```
Comment (robscott) [inline process/process.go]:
```suggestion
	// initialize a goroutine to process row data from inputQueue and
	// put the processed data into another queue to handle results
```

File: process/process.go [hunk 35]
```diff
@@ -31,18 +30,24 @@ const inZoneTrafficScoreWeight, deviationScoreWeight, sliceScoreWeight = 0.4, 0.
...
+	outputQueue, err := startSimulation(alg, inputQueue)
 	if err != nil {
 		return err
 	}
 
-	err = parseResult(outputFile, outputArray)
-	return err
+	// parse results from outputQueue and write to output file in the main
+	// thread which will end after all data has been processed
```
Comment (robscott) [inline process/process.go]:
```suggestion
	// parse results from outputQueue and write to output file
```

File: process/process.go [hunk 36]
```diff
@@ -66,33 +71,49 @@ type outputData struct {
...
-		if err != nil {
-			return outputArray, fmt.Errorf("error updating region for input : %s, %v", rowData.name, err)
-		}
-		simRes, err := model.StartSimulation()
-		if err != nil {
-			return outputArray, fmt.Errorf("error starting simulation for input : %s, %v", rowData.name, err)
+	outputQueue := make(chan outputData)
+	// since we only have one goroutine working on the data processing, although
+	// the model is not thread safe, for simplicity we create the model outside
+	// the goroutine but use it inside the goroutine.
```
Comment (robscott) [inline process/process.go]:
```suggestion
	// Some simplifications here result in this code not being threadsafe.
	// Do not use more than one goroutine to process this queue.
```

**Review by robscott — COMMENTED** [review]
Mostly LGTM, just a few minor nits around wording.

File: process/process.go (hunk 36 above)
Comment (ckyiricky) [reply process/process.go]: changed

File: process/process.go (hunk 35 above)
Comment (ckyiricky) [reply process/process.go]: changed

File: process/process.go (hunk 34 above)
Comment (ckyiricky) [reply process/process.go]: changed

File: process/process.go (hunk 33 above)
Comment (ckyiricky) [reply process/process.go]: I guess you meant 'goroutine to read xx and put' same as comments below. 
changed to 
'''
// initialize a goroutine to read row data from input file and
// put the converted row data into a queue
'''

File: process/input-parser.go (hunk 32 above)
Comment (ckyiricky) [reply process/input-parser.go]: fixed

**Comment by ckyiricky** [comment]
Fix typos

**Review by robscott — APPROVED** [review]
Thanks for the work on this!
